Issue339 try6 implement multi-broker subscriptions

petersilva opened the pull request:
close #339

Multiple subscriptions from a single broker work now.

This builds on the subscriptions that were built over the last few months.
The flowcb/gather/message plugin implementes one moth instance per subscription.

there is the *subscribe/dual_amis.conf* example added to the included ones in the distribution:

(available by: *sr3 add subscribe/dual_amis.conf* )
```
broker amqps://dd.weather.gc.ca/
topicPrefix v02.post

logReject on

# instances: number of downloading processes to run at once.  defaults to 1. Not enough for this case
instances 5

# expire, in operational use, should be longer than longest expected interruption
expire 10m

# new topic... in 2025
subtopic *.WXO-DD.bulletins.alphanumeric.#

broker amqps://hpfx.collab.science.gc.ca/

subtopic *.WXO-DD.bulletins.alphanumeric.#

#just dump into the current directory.
mirror false
directory /tmp/hpfx_amis/

```

can see that messages about the same item are received from both brokers.
The second one received is rejected as a duplicate, while the first one received
is acted on as a download.  Log extract below:

```
 ...
2025-04-07 23:37:52,034 [INFO] 3048061 sarracenia.flowcb.log after_accept accepted: (lag: 155.76 )  exchange: xpublic subtopic: 20250408.WXO-DD.bulletins.alphanumeric.20250408.SR.KWAL.03 a file with baseUrl: https://dd2.weather.gc.ca/ relPath: 20250408/WXO -DD/bulletins/alphanumeric/20250408/SR/KWAL/03/SRCN40_KWAL_080334___39084 sundew_extension: cvt_nws_bulletins-sr:KWAL:SR:3:Direct:20250408033514 id: r9wR/s5 size: 180 
 ...

025-04-07 23:37:52,046 [INFO] 3048061 sarracenia.flowcb.log after_accept accepted: (lag: 155.76 )  exchange: xpublic subtopic: 20250408.WXO-DD.bulletins.alphanumeric.20250408.SR.KWAL.03 a file with baseUrl: https://hpfx.collab.science.gc.ca relPath: /20250408/WXO-DD/bulletins/alphanumeric/20250408/SR/KWAL/03/SRCN40_KWAL_080334___39084 sundew_extension: cvt_nws_bulletins-sr:KWAL:SR:3:Direct:20250408033514 id: r9wR/s5 size: 180 
...
2025-04-07 23:37:54,007 [INFO] 3048061 sarracenia.flowcb.log after_work rejected: 406 mtime not newer /tmp/hpfx_amis/SRCN40_KWAL_080334___39084 
...

2025-04-07 23:37:54,007 [INFO] 3048061 sarracenia.flowcb.log after_work downloaded ok: /tmp/hpfx_amis/SRCN40_KWAL_080334___39084 

```

so this works, and the flow tests seem to work.

petersilva committed:
subscription broker was getting incorrectly converted to string as side-effect of write

petersilva committed:
remove url validator moved to credential class

andreleblanc11 committed:
Email sender bug when file type is None + logging improvements

petersilva committed:
use subscription instead  of settings

petersilva committed:
fixing auto_delete

petersilva committed:
no error when subscription.json does not exist

petersilva committed:
make cleanup use subscriptions

petersilva committed:
convert sr3 declare to use subscriptions

petersilva committed:
bug when opt does not contain messageAgeMax found in sr3 cleanup

petersilva committed:
fix #339 implementing multi broker support in gather

petersilva committed:
fix cleanup failure

petersilva committed:
for multi-broker #339 fixing metrics

petersilva committed:
fixing ack to match AMQP re-factor a few months ago

petersilva committed:
multi-broker compatibility #339 for mqtt, passes static

petersilva committed:
mqtt dynamic now passes

petersilva committed:
example of dual subscription

petersilva committed:
example of dual subscription: forgot logReject

andreleblanc11 commented:
I think I may have found a small problem when multiple **queue names** get defined in a config with multiple brokers. I think we would want support for multiple defined `queueName`s in a config?

I don't think this problem was detected in the test config you've referenced because the same queueName would've been used on both brokers?

This is the config I've used in my test.

```
broker amqps://dd.weather.gc.ca/
topicPrefix v02.post

# instances: number of downloading processes to run at once.  defaults to 1. Not enough for this case
instances 5

# expire, in operational use, should be longer than longest expected interruption
expire 10m

# new topic... in 2025
queueName q_anonymous.${CONFIG}.${HOSTNAME}.ANL-TEST
subtopic *.WXO-DD.bulletins.alphanumeric.#

broker amqps://hpfx.collab.science.gc.ca/

queueName q_anonymous.${CONFIG}.${HOSTNAME}.ANL-TEST01
subtopic *.WXO-DD.bulletins.alphanumeric.#

#just dump into the current directory.
mirror false
directory /net/local/home/leblanca/tmp/hpfx_amis/
accept .*
acceptUnmatched False
```

The queue on the HPFX broker that was created had `q_anonymous.issue339.hostname.ANL-TEST01` as its name, which is correct.

However, the queue name on the DD broker was also named `q_anonymous.issue339.hostname.ANL-TEST01`, which is wrong.

I think there might be some weirdness with the `*qname` file, because there is only one that's created? 
Maybe there's someplace that's still referencing `self.o.['queueName']` or something of the sorts.

```
ls subscribe/issue339/subscri*
subscribe/issue339/subscribe.issue339.anonymous.qname  subscribe/issue339/subscribe_issue339_02.pid  subscribe/issue339/subscribe_issue339_04.pid  subscribe/issue339/subscriptions.json
subscribe/issue339/subscribe_issue339_01.pid           subscribe/issue339/subscribe_issue339_03.pid  subscribe/issue339/subscribe_issue339_05.pid


cat ~/.cache/sr3/subscribe/issue339/subscriptions.json
[{"broker": "amqps://anonymous@dd.weather.gc.ca/", "bindings": [{"exchange": "xpublic", "prefix": ["v02", "post"], "sub": ["*.WXO-DD.bulletins.alphanumeric.#"]}], "queue": {"name": "q_anonymous.issue339.hostname.ANL-TEST01", "cleanup_needed": null, "auto_delete": false, "durable": true, "expire": 600.0, "prefetch": 25, "bind": true, "declare": true, "tlsRigour": "normal"}}, {"broker": "amqps://anonymous@hpfx.collab.science.gc.ca/", "bindings": [{"exchange": "xpublic", "prefix": ["v02", "post"], "sub": ["*.WXO-DD.bulletins.alphanumeric.#"]}], "queue": {"name": "q_anonymous.issue339.hostname.ANL-TEST01", "cleanup_needed": null, "auto_delete": false, "durable": true, "expire": 600.0, "prefetch": 25, "bind": true, "declare": true, "tlsRigour": "normal"}}]

cat ~/.cache/sr3/subscribe/issue339/subscribe.issue339.anonymous.qname
q_anonymous.issue339.hostname.ANL-TEST01
```

petersilva commented:
To be clear, this does not close the issue. It's just that it has enough progress to be worth merging.
Having multiple brokers in a single config works properly when everything is fine.

Further work needed before #339 is really done:

* test what happens when one broker fails... need to adjust things so that we keep using the other broker, and only check once in a while on the one that is down... and not wait too much for it.

* Review how  *broker* and *subtopic* interact in the grammar... maybe make it easier... maybe not.  See #339 for more detail.

petersilva commented:
@andreleblanc11 I reproduced the queue naming problem.  working on it.

petersilva committed:
config dump print broker in old_subscriptions

petersilva commented:
Still looking at the qname thing.   I have reproduced it once in a while, but not reliably.  It seems to mostly work... but occasionally not.

Looking at .qname file is not productive/helpful.  look at subscriptions.json instead.
If there is a problem in that file (incorrect queue names) then it is a real problem... I've found that once, but now trying to reproduce it, everything is working fine... frustrating.

Please note that the .qname file is going away... it's just a vestige of how state was stored before "subscriptions.json" came along.  It is only still maintained to allow for clean upgrades. at some point, support for qname will disappear.  it does not, and will not support multiple subscriptions.  

The .qname file is not supposed to be used other than when no subscription exists (the upgrade mechanism.) and the file is only written to allow people to downgrade seamlessly with the 100% of existing configurations that have a single queue.  There is no way for .qname files to work correctly with multiple queues because it is missing a lot of information.  That's why it is being replaced by subscriptions.json. So examining .qname in the case where multiple queues are defined is not a failure mode.

After a few more versions, .qname file support will be removed.

petersilva commented:
OK, the failure mode I actually found:
* two identical *broker* statements (not different.), but with different queue names.

result should be two subscriptions, one per queue, but instead, it is one subscription with only one of the queues, in it.

petersilva commented:
@andreleblanc11  can you check if your case actually works fine? I think the content of subscriptions.json should have the two queue names, and the brokers should have the correct queue names... but I'm not sure.

petersilva commented:
multiple queues with the same broker is a genuine design limitation.  interesting.

petersilva committed:
add *template* to subscription, to be able to match old_subscriptions.

had the problem that if someone declares a second queue on the same
broker, when reading from the subscription, it would match on the
broker... so the two subscriptions would collapse into one.

To fix that, created a "template" (the value of the queueName option
as entered in the config) as an additional matching parameter.  Now
one can have multiple queues on the same broker without risk of
confusion.

petersilva commented:
OK...
* changed when *queueName* and *queueShare* are evaluated to keep them as-input for a lot longer... so the values can be stored in subscriptions.json.  as entered (before any substitutions)  as a queue *template*
* when looking for existing queues, need to look them up by *broker* and a *template* that matches the current value of queueName.   

That change necessitated *queuename_resolved* being added to the options structure... actually not sure if that's needed, it might be ok with just a local variable... but put it there for now.  It works at least...  subject to future refinement.

OK that's pushed... Please try again.

petersilva commented:
something weird with queuename evaluation... numbers not matching...

andreleblanc11 commented:
I pulled in your latest commit to my test setup and tried again.

I haven't changed the configuration in my test case.

```
broker amqps://dd.weather.gc.ca/
topicPrefix v02.post

# instances: number of downloading processes to run at once.  defaults to 1. Not enough for this case
instances 5

# expire, in operational use, should be longer than longest expected interruption
expire 10m

# new topic... in 2025
queueName q_anonymous.${CONFIG}.${HOSTNAME}.ANL-TEST
subtopic *.WXO-DD.bulletins.alphanumeric.#

broker amqps://hpfx.collab.science.gc.ca/

queueName q_anonymous.${CONFIG}.${HOSTNAME}.ANL-TEST01
subtopic *.WXO-DD.bulletins.alphanumeric.#

#just dump into the current directory.
mirror false
directory /net/local/home/leblanca/tmp/hpfx_amis/
accept .*
acceptUnmatched False
```

I'm still seeing the same `queueName` (`q_anonymous.issue339.hostname.ANL-TEST01`) on both `hpfx.collab.science.gc.ca` and on `dd.weather.gc.ca` brokers. 

It seems to happen to me every time that I try with my test case. Here is the content of the `subscribptions.json`. 
We see the `queueName` is the same for both brokers.

```
cat subscriptions.json
[{"broker": "amqps://anonymous@dd.weather.gc.ca/", "bindings": [{"exchange": "xpublic", "prefix": ["v02", "post"], "sub": ["*.WXO-DD.bulletins.alphanumeric.#"]}], "queue": {"name": "q_anonymous.issue339.hostname.ANL-TEST01", "cleanup_needed": null, "auto_delete": false, "durable": true, "expire": 600.0, "prefetch": 25, "bind": true, "declare": true, "tlsRigour": "normal"}}, {"broker": "amqps://anonymous@hpfx.collab.science.gc.ca/", "bindings": [{"exchange": "xpublic", "prefix": ["v02", "post"], "sub": ["*.WXO-DD.bulletins.alphanumeric.#"]}], "queue": {"name": "q_anonymous.issue339.hostname.ANL-TEST01", "cleanup_needed": null, "auto_delete": false, "durable": true, "expire": 600.0, "prefetch": 25, "bind": true, "declare": true, "tlsRigour": "normal"}}]
```

I tried using the same broker, `amqps://anonymous@dd.weather.gc.ca/`  for each assigned queue, and I get a similar result, only one queue name gets defined. Here is the subscriptions.json file for this case.

```
cat subscriptions.json
[{"broker": "amqps://anonymous@dd.weather.gc.ca/", "bindings": [{"exchange": "xpublic", "prefix": ["v02", "post"], "sub": ["*.WXO-DD.bulletins.alphanumeric.#"]}], "queue": {"name": "q_anonymous.issue339.hostname.ANL-TEST01", "cleanup_needed": null, "auto_delete": false, "durable": true, "expire": 600.0, "prefetch": 25, "bind": true, "declare": true, "tlsRigour": "normal"}}, {"broker": "amqps://anonymous@dd.weather.gc.ca/", "bindings": [{"exchange": "xpublic", "prefix": ["v02", "post"], "sub": ["*.WXO-DD.bulletins.alphanumeric.#"]}], "queue": {"name": "q_anonymous.issue339.hostname.ANL-TEST01", "cleanup_needed": null, "auto_delete": false, "durable": true, "expire": 600.0, "prefetch": 25, "bind": true, "declare": true, "tlsRigour": "normal"}}]
```

petersilva commented:
have you done  *cleanup* between attempts?   the subscriptions.json maintenance has been updated... need a currently written one.

andreleblanc11 commented:
Sorry that was before your latest changes. Let me try again.

andreleblanc11 commented:
Each time I restart my test case, I do a `sr3 stop subscribe/my-test-config && sr3 cleanup subscribe/my-test-config`

I pulled in your most recent commit, reinstalled it on my machine, performed the commands above, and tried to start it back up, but now I'm getting an error

```
sr3 start subscribe/my-test-config

Traceback (most recent call last):
  File "/net/local/home/leblanca/.local/bin/sr3", line 8, in <module>
    sys.exit(main())
  File "/net/local/home/leblanca/.local/lib/python3.10/site-packages/sarracenia/sr.py", line 3204, in main
    gs = sr_GlobalState(cfg, cfg.configurations)
  File "/net/local/home/leblanca/.local/lib/python3.10/site-packages/sarracenia/sr.py", line 1348, in __init__
    self._read_configs()
  File "/net/local/home/leblanca/.local/lib/python3.10/site-packages/sarracenia/sr.py", line 334, in _read_configs
    cfgbody.parse_file(cfg,c)
  File "/net/local/home/leblanca/.local/lib/python3.10/site-packages/sarracenia/config/__init__.py", line 1554, in parse_file
    self.parse_line( component, cfg, cfname, lineno, l.strip() )
  File "/net/local/home/leblanca/.local/lib/python3.10/site-packages/sarracenia/config/__init__.py", line 1668, in parse_line
    self._parse_binding(v)
  File "/net/local/home/leblanca/.local/lib/python3.10/site-packages/sarracenia/config/__init__.py", line 1395, in _parse_binding
    self.resolved_queueName = self._resolveQueueName(self.component,self.config)
  File "/net/local/home/leblanca/.local/lib/python3.10/site-packages/sarracenia/config/__init__.py", line 1870, in _resolveQueueName
    if self.broker == s['broker'] and self.queueName == s['queue']['template']:
KeyError: 'template'
```

I think it might be a problem with the `subscriptions.json` not existing yet? Not sure

```
leblanca@hostname:~/.cache/sr3/subscribe/my-test-config$ ls -al
total 8
drwxr-xr-x  2 leblanca leblanca 4096 Apr  9 15:11 .
drwxr-xr-x 11 leblanca leblanca 4096 Apr  8 13:17 ..
```

petersilva committed:
better timing for evaluation for more consisten random id selection in queue names

petersilva committed:
refactor to use subscriptions to store resolved queue names

queueName becomes a pure setting:
* never overwritten.
* has a logical default (used to be None?, but default added in code.)

resolved_queueName becomes a local variable, with storage in the
subscriptions structure.

Ensure first subscription is stored in .qname file. qname file is
only for compatibility with older versions.  It is never read
when subscriptions are available.

petersilva committed:
make .qname very explicitly a fallback when no subscriptions available

petersilva committed:
some attributes are not properly named in subscriptions.

petersilva committed:
remove queuefile_read setting. does nothing.

petersilva commented:
ok... so the actual problem was that I was comparing:

when looking in the subscriptions for queues whose names have already been resolved, need to look for the same broker, and the same queue template.  To check the broker I used:  *self.broker == s['broker']*   and they were coming out different when the broker values look the same to me.  I don't understand what was wrong with that.  I wrapped each side with *str()* and now the comparisons work properly. odd.

Did a whole pile of irrelevant but good anyways refactors along the way.

not done yet... still working on it.

petersilva committed:
return empty list for one, not None

petersilva committed:
look at old subscriptions first, then new ones

petersilva committed:
queueName and queueShare must be expanded very late.

petersilva committed:
remove debug output from qname parsing

petersilva committed:
somehow the broker stringifying disappeared... need it back

petersilva commented:
ok @andreleblanc11  you can try again. I think everything works now.  One thing I have not tested is upward/downward compatibility.  I will do that... but It's probably fine.

petersilva commented:
upward compat.. bad... wait.

petersilva committed:
fixed upgrade from existing subscription files (<=3.0.58)

petersilva commented:
ok upward compatibility looks good now.  ok to try.

petersilva commented:
don't approve yet ... worth testing... but don't approve yet... I want to look at unit tests.

petersilva committed:
adjusting examples to use subscriptions.

All kinds of sadness here (left to fix later?)

why need expire in queue defn and global options (that override queue
options.)

removed some debug prints.

flow_api_consumer ... messageCountMax 5 results in 6 messages...
that was erroring... changed the test to expect 6 for now.

petersilva committed:
get unit tests to pass after multi subcriptions

petersilva commented:
ok unit testing and flow_maint_test (aka examples) all pass for me now.  I think it's good to go.
We might want more elaborate unit tests later, but what is here is mergeable.

reidsunderland committed:
copy multi-broker changes to amqp consumer

reidsunderland commented:
I added a commit to update amqp consumer as well. We're going to test on dev

reidsunderland committed:
rename queue variable, to avoid conflict with module

reidsunderland committed:
fix problems related to changes for non-blocking?

petersilva commented:
@reidsunderland's changes look good!  I was not worrying about non-blocking yet... wanted to get the everything is OK stuff merged first... I dunno... we might need similar changes in the moth/amqp.py as well.  but It does not have to be in this PR.

andreleblanc11 commented:
The multiple queue names w/ the multiple brokers does seem to work now.

For 

```
broker amqps://dd.weather.gc.ca/
topicPrefix v02.post

# instances: number of downloading processes to run at once.  defaults to 1. Not enough for this case
instances 5

# expire, in operational use, should be longer than longest expected interruption
expire 10m

# new topic... in 2025
queueName q_anonymous.${CONFIG}.${HOSTNAME}.ANL-TEST
subtopic *.WXO-DD.bulletins.alphanumeric.#

broker amqps://hpfx.collab.science.gc.ca/
queueName q_anonymous.${CONFIG}.${HOSTNAME}.ANL-TEST01
subtopic *.WXO-DD.bulletins.alphanumeric.#

#just dump into the current directory.
mirror false
directory /net/local/home/leblanca/tmp/hpfx_amis/
accept .*
acceptUnmatched False
```

I'm going to try to add more queue options (i.e. prefetch, expire, etc.) to the separate queues to see if I can break some more things.

reidsunderland commented:
I think the non-blocking changes for amqp consumer have actually been needed since the earlier refactor (already merged) that removed the long sleeps from the amqp class. I just didn't notice the problem until now

petersilva commented:
note: 
* the maintenance tests fail because they retrieve examples from the development branch, which won't work with this branch.  When I run on 24.04 dev box, with local code, they pass.  So while they fail now, they will start passing after merge.
* similar thing for unit tests... I think... not sure about that one.

andreleblanc11 commented:
I've tested a couple of things with the AMQP and the AMQP consumer protocols.
- Different `expire` and `prefetch` options for each broker connection
- Different `subtopic`s for each broker connection.
 
I couldn't find any problems with all of these use cases. I've also tried adding a third broker in the configuration (just to see what would happen) and I was able to connect to it no problem.

Here is what my test configuration looks like after my tests.

```
amqp_consumer True

broker amqps://dd.weather.gc.ca/
topicPrefix v02.post

# instances: number of downloading processes to run at once.  defaults to 1. Not enough for this case
instances 5

# expire, in operational use, should be longer than longest expected interruption
expire 10m
prefetch 25

# new topic... in 2025
queueName q_anonymous.${CONFIG}.${HOSTNAME}.ANL-TEST
subtopic *.WXO-DD.bulletins.alphanumeric.#


broker amqps://hpfx.collab.science.gc.ca/

expire 30m
prefetch 1
queueName q_anonymous.${CONFIG}.${HOSTNAME}.ANL-TEST01
subtopic *.WXO-DD.observations.#
subtopic *.WXO-DD.bulletins.alphanumeric.#

broker amqp://tfeed@localhost/

queueName q_anonymous.${CONFIG}.${HOSTNAME}.ANL-TEST02
subtopic #

#just dump into the current directory.
mirror True
directory /net/local/home/leblanca/tmp/hpfx_amis/
accept .*
acceptUnmatched False

debug True
logReject on
```

Subscriptions file

```
[{"broker": "amqps://anonymous@dd.weather.gc.ca/", "bindings": [{"exchange": "xpublic", "prefix": ["v02", "post"], "sub": ["*.WXO-DD.bulletins.alphanumeric.#"]}], "queue": {"name": "q_anonymous.issue339.edcm-dirt22-3.edc-mtl.ec.gc.ca.ANL-TEST", "template": "q_anonymous.${CONFIG}.${HOSTNAME}.ANL-TEST", "cleanup_needed": null, "bind": true, "declare": true, "auto_delete": false, "durable": true, "expire": 600.0, "prefetch": 25, "tlsRigour": "normal"}}, {"broker": "amqps://anonymous@hpfx.collab.science.gc.ca/", "bindings": [{"exchange": "xpublic", "prefix": ["v02", "post"], "sub": ["*.WXO-DD.observations.#"]}, {"exchange": "xpublic", "prefix": ["v02", "post"], "sub": ["*.WXO-DD.bulletins.alphanumeric.#"]}], "queue": {"name": "q_anonymous.issue339.edcm-dirt22-3.edc-mtl.ec.gc.ca.ANL-TEST01", "template": "q_anonymous.${CONFIG}.${HOSTNAME}.ANL-TEST01", "cleanup_needed": null, "bind": true, "declare": true, "auto_delete": false, "durable": true, "expire": 1800.0, "prefetch": 1, "tlsRigour": "normal"}}, {"broker": "amqp://tfeed@localhost/", "bindings": [{"exchange": "xpublic", "prefix": ["v02", "post"], "sub": ["#"]}], "queue": {"name": "q_anonymous.issue339.edcm-dirt22-3.edc-mtl.ec.gc.ca.ANL-TEST02", "template": "q_anonymous.${CONFIG}.${HOSTNAME}.ANL-TEST02", "cleanup_needed": null, "bind": true, "declare": true, "auto_delete": false, "durable": true, "expire": 1800.0, "prefetch": 1, "tlsRigour": "normal"}}]
```

petersilva committed:
document change from bindings to subscriptions

petersilva committed:
French updates corresponding to previous English ones

petersilva commented:
yeah, there is no limit to number of brokers, or number of queues. so now can implement prioritized feeds with a single subscriber using separate queues for high and normal priority.  Obviously we can do HA without shovels also... it should make a big difference once it gets shaken down.  Probably want to make new demonstrations... need a lot fewer winnows and shovels.

petersilva committed:
remove config.bindings, as they are replaced by config.subscriptions

petersilva committed:
say which broker subscription is broken, when failure happens

petersilva committed:
more information about amqp connection failures.

petersilva commented:
I just tested AMQP recover with multiple brokers, and it works perfectly already... yay!
MQTT I just looked at... and it looks quite wrong... future work.

petersilva committed:
MQTT fixes for subscriptions, mostly... but other small fixes.

- amqp.py should abort when stopping not wait until next retry
- more cases where say what broker in message, no longer obvious.
- MQTT subscribe parsing was wrong... (was still using global var not
  subscriptions.)
- added next_connect_time ebo connection spacing to MQTT.

petersilva commented:
I tested mqtt stuff and it is much better now.  there was no EBO to let other class instances keep running.
MQTT flow tests passed static and dynamic... flakey had 1 fail... didn't investigate.

andreleblanc11 commented:
I think I may have found a problem while running  `sr3 dump`? The current development branch seems unaffected. 

```
sarra@dev01:~/ sr3 dump
....
dump output 
....
        },
        "poll": {
Traceback (most recent call last):
  File "/local/home/sarra/.local/bin/sr3", line 11, in <module>
    load_entry_point('metpx-sr3', 'console_scripts', 'sr3')()
  File "/local/home/sarra/sr3/sarracenia/sr.py", line 3235, in main
    gs.dump()
  File "/local/home/sarra/sr3/sarracenia/sr.py", line 2535, in dump
    print('\t\t\"%s\" :  %s ' % (cfg, json.dumps(self.states[c][cfg])))
  File "/usr/lib/python3.6/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.6/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.6/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python3.6/json/encoder.py", line 180, in default
    o.__class__.__name__)
TypeError: Object of type 'Credential' is not JSON serializable
```

petersilva commented:
reproduced... yup. hang on.

petersilva committed:
make credentials serializable to JSON encoders

petersilva commented:
@andreleblanc11 fixed! git pull and try again!

andreleblanc11 commented:
Looks good now :smile:

petersilva committed:
fix #1420 precise time comparison in nodupe/disk engine

petersilva committed:
fix #1420 more compactly and correctly

petersilva commented:
ok, well somebody needs to bless it if there aren't any objections anymore.

petersilva commented:
I just ran the no_mirror flow test on this branch with MQTT and it also passes now... but not immediately.  the sender is a bit slow, and I have to run flow_check.sh after a few minutes...

petersilva commented:
With alll the subsequent work done, I think #339 is now completely addressed by this PR.
so I changed the initial comment to get it closed when this gets merged.

andreleblanc11 committed:
Fix american bulletins not renaming properly + rand. number generations + interesting batch note

petersilva committed:
fix post_exchange parsing, it had a weirdness.

was adding the suffix multiple times. dunno why.

petersilva committed:
remove 20.04 from github workflows, no longer supported

petersilva commented:
the @andreleblanc11 commits are just me rebasing for what has been merged while this is pending.

petersilva commented:
bless soon?
